src/v/v8_engine/script_dispatcher.h

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "bytes/iobuf.h"
+#include "model/metadata.h"
+#include "utils/mutex.h"
+#include "v8_engine/script.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/util/variant_utils.hh>
+
+#include <absl/container/node_hash_map.h>
+#include <absl/hash/hash.h>
+
+#include <exception>
+#include <iostream>
+#include <variant>
+
+namespace v8_engine {
+
+struct script_compile_params {
+    ss::sstring function_name;
+    iobuf code;
+    size_t max_heap_size_bytes{10000};
+    size_t run_timeout_ms{200};
+};
+
+// This class implement database for v8_runtimes. It can be updated by
+// data_policy_manager. Fetch request will only run get for getting v8_runtimes
+template<typename Executor>
+class script_dispatcher {
+    constexpr static size_t _mutex_count{50};
+
+public:
+    explicit script_dispatcher(Executor& executor)
+      : _mutexes(_mutex_count)
+      , _executor(executor) {}
+
+    ss::future<>
+    update(model::topic_namespace& topic, script_compile_params script_params) {
+        auto hash = absl::Hash<model::topic_namespace>{}(topic);
+        auto lock = co_await _mutexes[hash % _mutexes.size()].get_units();
+        _scripts.insert_or_assign(topic, std::move(script_params));
+    }
+
+    ss::future<> remove(model::topic_namespace topic) {
+        auto hash = absl::Hash<model::topic_namespace>{}(topic);
+        auto lock = co_await _mutexes[hash % _mutexes.size()].get_units();
+        _scripts.erase(topic);
+    }
+
+    /// This method return v8_runtimes for data-policy (aka topic on curent
+    /// moment). If hash_table contains compile params get needs to compile this
+    /// script, so on this moment we need to lock hash_table. Also if sbmd try
+    /// to get v8_runtimes it needs to wait compilation and do not compile twice
+    ss::future<ss::lw_shared_ptr<script>> get(model::topic_namespace topic) {
+        auto hash = absl::Hash<model::topic_namespace>{}(topic);
+        auto lock = co_await _mutexes[hash % _mutexes.size()].get_units();
```

Comment:
You can remove update & remove as you can guarantee up until the first async call in this method, since there are no scheduling points, you won't have concurrent access issues with this data structure `_mutexes`

Reply:
Problem is, get can compile and do async work, it can be problem is smbd change data_policy in compile moment, and get rewrite the new information when set runtime

Reply:
As long as the internal data isn't shared, if any entity needs to come and hit this structure, it should encounter the mutex and wait. Once the compile is finished, an update can occur which can then initiate a new event, such as a new compile.

Reply:
Oh, lol. This part with mutexes is completely wrong...
Sorry